feat(wallet_ffi): add coin join and split

agubarev opened the pull request:
Description
---
Adding new FFI `wallet_coin_join()` function and working on unit tests for `wallet_coin_join()` and an already existing `wallet_coin_split()`

Motivation and Context
---
To enable users to join (collapse) UTXOs depending on the selected strategy, with default behavior to join outputs selected by the user.

How Has This Been Tested?
---
unit tests (WIP)

agubarev committed:
added FFI `wallet_get_utxos()`, `destroy_tari_outputs` + unit test
added a simple output querying function

Signed-off-by: Andrey Gubarev <agubarev@protonmail.com>

agubarev committed:
added FFI `wallet_get_utxos()`, `destroy_tari_outputs` + unit test.
added a simple output querying function.
added an output database shortcut for easier access from the wallet context, circumventing hops through async fall-through constructs; updated all usages - `cargo test` passing.

Signed-off-by: Andrey Gubarev <agubarev@protonmail.com>

agubarev committed:
added coin join + initial unit test (still WIP)

stringhandler commented on base_layer/wallet/src/output_manager_service/handle.rs:
```diff
@@ -101,10 +105,16 @@ pub enum OutputManagerRequest {
+    CreateCoinJoin {
+        target_amount: MicroTari,
+        fee_per_gram: MicroTari,
+        commitments: Option<Vec<PublicKey>>,
```
Rather use an empty vec to represent "None"
```suggestion
        commitments: Vec<PublicKey>,
```

stringhandler commented on base_layer/wallet/src/output_manager_service/handle.rs:
```diff
@@ -670,6 +699,26 @@ impl OutputManagerHandle {
+        &mut self,
+        target_amount: MicroTari,
+        fee_per_gram: MicroTari,
+        commitments: Option<Vec<PublicKey>>,
```
```suggestion
        commitments: Vec<PublicKey>,
```

agubarev replied:
Done

agubarev committed:
Update base_layer/wallet/src/output_manager_service/handle.rs

Co-authored-by: stringhandler <stringhandler@gmail.com>

agubarev committed:
Update base_layer/wallet/src/output_manager_service/handle.rs

Co-authored-by: stringhandler <stringhandler@gmail.com>

agubarev committed:
adding TariVector to substitute TariOutputs et al (WIP)

sdbondi commented on base_layer/wallet_ffi/src/lib.rs:
```diff
@@ -218,20 +230,175 @@ pub struct TariWallet {
+        })
+    }
+
+    fn to_string_vec(&self) -> Result<Vec<String>, InterfaceError> {
```
The `to_*` functions should be marked as unsafe because they do not provide any additional safety guarantees than from_raw_parts (which is unsafe). If the data that TariVector points to is null/deallocated the Vec<String> will be invalid and the program will crash. Moreover if the vector is mutated (as you are free to do with an owned Vec, it could be reallocated and the pointer is no longer valid.

~~I think the method signature should be:
`unsafe fn as_string_slice(&self) -> Result<&[&str], InterfaceError>`~~ wont work

agubarev committed:
rewrote coin_split function
fixed TariVector conversion bug
found a problem in the coin_join approach, need reconsideration

this commit is for review only before the final brushing/wrap-up

Signed-off-by: Andrei Gubarev <agubarev@users.noreply.github.com>

stringhandler reviewed (dismissed):
I would have preferred less duplication of code, but I'm not sure how often this functionality is actually going to be used, so let's try it and we can DRY  it up later

agubarev committed:
FFI join & split functionality + tests

Signed-off-by: Andrei Gubarev <agubarev@users.noreply.github.com>

agubarev committed:
bugfixes, testfixes, cleanup, refactored select_utxos()

Signed-off-by: Andrei Gubarev <agubarev@users.noreply.github.com>

agubarev committed:
bugfixes, testfixes, cleanup, refactored select_utxos()

Signed-off-by: Andrei Gubarev <agubarev@users.noreply.github.com>
